activesupport/lib/active_support/cache.rb

```diff
@@ -955,16 +962,31 @@ def validate_options(options)
           options
         end
 
-        # Expands and namespaces the cache key.
+        # Expands, namespaces and truncates the cache key.
         # Raises an exception when the key is +nil+ or an empty string.
         # May be overridden by cache stores to do additional normalization.
         def normalize_key(key, options = nil)
+          key = expand_and_namespace_key(key, options)
+          truncate_key(key)
+        end
+
+        def expand_and_namespace_key(key, options = nil)
           str_key = expanded_key(key)
           raise(ArgumentError, "key cannot be blank") if !str_key || str_key.empty?
 
           namespace_key str_key, options
         end
 
+        def truncate_key(key)
+          if key && @max_key_size && key.bytesize > @max_key_size
+            suffix = ":hash:#{ActiveSupport::Digest.hexdigest(key)}"
+            truncate_at = @max_key_size - suffix.bytesize
+            "#{key.byteslice(0, truncate_at)}#{suffix}"
```

Comment:
```suggestion
            key = key.byteslice(0, truncate_at)
            key.scrub!("")
            "#{key}#{suffix}"
```

If you need a test string, I did this:

```ruby
key = "a" * 25 + "€" + "b" * 50
```